kerchunk/hdf.py

```diff
@@ -89,8 +94,23 @@ def __init__(
             fs, path = fsspec.core.url_to_fs(h5f, **(storage_options or {}))
             self.input_file = fs.open(path, "rb")
             url = h5f
-        else:
+            self._h5f = h5py.File(self.input_file, mode="r")
+        elif isinstance(h5f, io.IOBase):
             self.input_file = h5f
+            self._h5f = h5py.File(self.input_file, mode="r")
+        elif isinstance(h5f, (h5py.File, h5py.Group)):
+            # assume h5py object (File or group/dataset)
+            self._h5f = h5f
+            fs, path = fsspec.core.url_to_fs(url, **(storage_options or {}))
+            self.input_file = fs.open(path, "rb")
+        elif isinstance(h5f, h5py.Dataset):
+            group = h5f.file.create_group(f"{h5f.name} ")
+            group[h5f.name] = h5f
```

Comment:
this barfs, unfortunately:
```
activestorage/netcdf_to_zarr.py:46: in gen_json
    h5chunks = SingleHdf5ToZarr(_dataset, file_url,
../miniconda3/envs/pyactive/lib/python3.12/site-packages/kerchunk/hdf.py:108: in __init__
    group[h5f.name] = h5f
../miniconda3/envs/pyactive/lib/python3.12/site-packages/h5py/_hl/group.py:468: in __setitem__
    h5o.link(obj.id, self.id, name, lcpl=lcpl, lapl=self._lapl)
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   OSError: Unable to create link (name already exists)

h5py/h5o.pyx:201: OSError
```

Reply:
sorry, forgot to mention how I am creating the call:
```python
    elif storage_type == "s3" and storage_options is not None:
        storage_options = storage_options.copy()
        storage_options['default_fill_cache'] = False
        storage_options['default_cache_type'] = "first"
        fs = s3fs.S3FileSystem(**storage_options)
        fs2 = fsspec.filesystem('')
        tk1 = time.time()
        with fs.open(file_url, 'rb') as s3file:
            _file = h5py.File(s3file, mode="w")
            _dataset = _file[varname]
            h5chunks = SingleHdf5ToZarr(_dataset, file_url,
                                        inline_threshold=0)
 ```

Reply:
This tried to rewrite the file??
I thought there would be a way to update the in-memory version without changing the file at all. If that's not true, it leaves us in a pickle, since remote files can't be rewritten ever (without copying to local, which we don't want).

Reply:
AFAIK and from my tests it's not trying to write to file, the only way it nee allows group creation is if the file object us opened in write mode - have not seen any actual fata transfers or writes though, it's just an annoyance that it won't allow new groups with existing dataset names

Reply:
So hacking the name fixes this?

Reply:
Yeah, just giving it any name that doesn't already exist, could be `cow-in-field` for the matter 😁

Reply:
OK, I did that, let me know what happens.

Reply:
Cheers! This should work now, I'll test on Monday. HDF5 is really strict with its names and such - probably bc it's a fairly thin border between a Dataset and a Group, but then again, they should support similar API's and methods on both